File: supabase/functions/_shared/monetization/referral_receipt.ts
// Canonical referral receipts, shared byte-for-byte with the Flutter client.
//
// Canonical JSON here: object keys sorted by code point at every level, no
// whitespace, integers only, ASCII-only strings. Every value in a receipt is
// an ASCII identifier, timestamp or enum, so a value outside that profile is
// refused rather than escaped differently by two JSON libraries. The fixture
// test/fixtures/monetization/referral_receipt.v1.json pins the exact bytes.

import { sha256Hex } from "./billing_crypto.ts";

export class InvalidReceipt extends Error {}

type Json = null | boolean | number | string | Json[] | { [key: string]: Json };

export function canonicalJson(value: unknown): string {
  if (value === null || typeof value === "boolean") {
    return JSON.stringify(value);
  }
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) throw new InvalidReceipt("number");
    return String(value);
  }
  if (typeof value === "string") {
    if (!/^[\x20-\x7e]*$/.test(value)) throw new InvalidReceipt("string");
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(",")}]`;
  if (typeof value === "object") {
    const entries = Object.keys(value as object).sort().map((key) => {
      if (!/^[a-z_]+$/.test(key)) throw new InvalidReceipt("key");
      return `${JSON.stringify(key)}:${
        canonicalJson((value as Record<string, unknown>)[key])
      }`;
    });
    return `{${entries.join(",")}}`;
  }
  throw new InvalidReceipt("type");
}

const uuid = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;
// UTC, second or millisecond precision, as Dart's toIso8601String() on a UTC
// DateTime truncated to milliseconds produces.
const timestamp = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{3})?Z$/;
const interactions = new Set([
  "teaching_acknowledged",
  "answered_correctly",
  "answered_incorrectly",
  "skipped",
]);
const receiptKeys = [
  "attempt_id",
  "campaign_id",
  "claim_id",
  "completed_at_client",
  "content_revision",
  "initial_coverage",
  "lesson_id",
  "schema_version",
  "started_at_client",
];

export interface ReferralReceipt {
  schema_version: 1;
  claim_id: string;
  campaign_id: string;
  content_revision: number;
  lesson_id: number;
  attempt_id: string;
  started_at_client: string;
  completed_at_client: string;
  initial_coverage: { exercise_id: number; interaction: string }[];
}

const isObject = (v: unknown): v is Record<string, unknown> =>
  typeof v === "object" && v !== null && !Array.isArray(v);

/**
 * Checks the receipt's shape and returns it with coverage sorted by exercise.
 * Manifest membership and ownership are the database's to decide.
 */
export function normalizeReceipt(raw: unknown): ReferralReceipt {
  if (
    !isObject(raw) ||
    Object.keys(raw).sort().join() !== receiptKeys.join() ||
    raw.schema_version !== 1 ||
    typeof raw.claim_id !== "string" || !uuid.test(raw.claim_id) ||
    typeof raw.attempt_id !== "string" || !uuid.test(raw.attempt_id) ||
    typeof raw.campaign_id !== "string" ||
    !/^[a-z0-9-]{1,64}$/.test(raw.campaign_id) ||
    !Number.isSafeInteger(raw.content_revision) ||
    !Number.isSafeInteger(raw.lesson_id) ||
    typeof raw.started_at_client !== "string" ||
    !timestamp.test(raw.started_at_client) ||
    typeof raw.completed_at_client !== "string" ||
    !timestamp.test(raw.completed_at_client) ||
    !Array.isArray(raw.initial_coverage) ||
    raw.initial_coverage.length === 0 || raw.initial_coverage.length > 200
  ) {
    throw new InvalidReceipt("shape");
  }
  const coverage = raw.initial_coverage.map((item) => {
    if (
      !isObject(item) || Object.keys(item).sort().join() !==
        "exercise_id,interaction" ||
      !Number.isSafeInteger(item.exercise_id) ||
      typeof item.interaction !== "string" ||
      !interactions.has(item.interaction)
    ) {
      throw new InvalidReceipt("coverage");
    }
    return {
      exercise_id: item.exercise_id as number,
      interaction: item.interaction,
    };
  }).sort((a, b) => a.exercise_id - b.exercise_id);
  return { ...(raw as unknown as ReferralReceipt), initial_coverage: coverage };
}

export const receiptDigest = (receipt: ReferralReceipt) =>
  sha256Hex(canonicalJson(receipt as unknown as Json));

/**
 * Play Integrity `requestHash`: binds the token to this account, claim,
 * campaign, receipt and single-use nonce. 64 hex characters, well inside
 * Play's 500-byte limit.
 */
export const integrityRequestHash = (input: {
  account_id: string;
  campaign_id: string;
  claim_id: string;
  nonce: string;
  receipt_digest: string;
}) => sha256Hex(canonicalJson(input));

File: lib/domain/entities/referral_receipt.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';

import 'enums.dart';
import 'exercise.dart';
import 'exercise_attempt_evidence.dart';
import 'exercise_outcome.dart';

/// The referral campaign whose manifest these receipts are checked against.
const referralCampaignId = 'a1-referral-v1';

/// Content revision the campaign's pinned manifest was reviewed for. A new
/// content revision needs a new reviewed manifest on the server first.
const referralContentRevision = 25;

/// How the learner first met one exercise in an attempt. Teaching cards are
/// acknowledged, never answered; the lesson player reports their Continue as
/// a skip, so that report is translated here rather than trusted.
enum ReferralInteraction {
  teachingAcknowledged('teaching_acknowledged'),
  answeredCorrectly('answered_correctly'),
  answeredIncorrectly('answered_incorrectly'),
  skipped('skipped');

  const ReferralInteraction(this.wireName);
  final String wireName;
}

/// Evidence that one lesson attempt was completed in the normal player: each
/// exercise's first interaction, the attempt, and the claim held when the
/// attempt began. It proves participation, not a unique human.
class ReferralReceipt {
  final String claimId;
  final int lessonId;
  final String attemptId;
  final DateTime startedAt;
  final DateTime completedAt;
  final Map<int, ReferralInteraction> coverage;

  const ReferralReceipt({
    required this.claimId,
    required this.lessonId,
    required this.attemptId,
    required this.startedAt,
    required this.completedAt,
    required this.coverage,
  });

  /// Builds coverage from the player's evidence: the first `initial` record
  /// per exercise. Returns null unless every exercise of the lesson has one,
  /// so a partial attempt never becomes a receipt.
  static ReferralReceipt? fromAttempt({
    required String claimId,
    required int lessonId,
    required String attemptId,
    required DateTime startedAt,
    required DateTime completedAt,
    required List<Exercise> exercises,
    required List<ExerciseAttemptEvidence> evidence,
  }) {
    final types = {for (final e in exercises) e.id: e.type};
    if (types.isEmpty) return null;
    final coverage = <int, ReferralInteraction>{};
    for (final record in evidence) {
      if (record.phase != ExerciseEvidencePhase.initial) continue;
      final type = types[record.exerciseId];
      if (type == null || coverage.containsKey(record.exerciseId)) continue;
      coverage[record.exerciseId] =
          type == ExerciseType.teaching
              ? ReferralInteraction.teachingAcknowledged
              : switch (record.outcome) {
                ExerciseOutcome.correct =>
                  ReferralInteraction.answeredCorrectly,
                ExerciseOutcome.incorrect =>
                  ReferralInteraction.answeredIncorrectly,
                ExerciseOutcome.skipped => ReferralInteraction.skipped,
              };
    }
    if (coverage.length != types.length) return null;
    return ReferralReceipt(
      claimId: claimId,
      lessonId: lessonId,
      attemptId: attemptId,
      startedAt: startedAt,
      completedAt: completedAt,
      coverage: coverage,
    );
  }

  /// The wire object. Coverage is sorted by exercise ID, as the server
  /// normalizes it, so both sides hash the same bytes.
  Map<String, Object> toJson() => {
    'schema_version': 1,
    'claim_id': claimId,
    'campaign_id': referralCampaignId,
    'content_revision': referralContentRevision,
    'lesson_id': lessonId,
    'attempt_id': attemptId,
    'started_at_client': referralTimestamp(startedAt),
    'completed_at_client': referralTimestamp(completedAt),
    'initial_coverage': [
      for (final id in coverage.keys.toList()..sort())
        {'exercise_id': id, 'interaction': coverage[id]!.wireName},
    ],
  };

  String get digest => sha256Hex(canonicalJson(toJson()));
}

/// UTC with exactly three fractional digits, the shape the server accepts.
String referralTimestamp(DateTime time) {
  final utc = time.toUtc();
  final truncated = DateTime.utc(
    utc.year,
    utc.month,
    utc.day,
    utc.hour,
    utc.minute,
    utc.second,
    utc.millisecond,
  );
  final iso = truncated.toIso8601String();
  // Dart omits the fraction when it is zero; always emit milliseconds.
  return iso.contains('.') ? iso : iso.replaceFirst('Z', '.000Z');
}

/// Canonical JSON shared with the server: keys sorted at every level, no
/// whitespace, integers only, printable-ASCII strings. Anything else throws
/// rather than risk a different encoding. Pinned by
/// test/fixtures/monetization/referral_receipt.v1.json.
String canonicalJson(Object? value) {
  if (value == null || value is bool) return jsonEncode(value);
  if (value is int) return value.toString();
  if (value is String) {
    if (!RegExp(r'^[\x20-\x7e]*$').hasMatch(value)) {
      throw const FormatException('Non-ASCII string in canonical JSON.');
    }
    return jsonEncode(value);
  }
  if (value is List) return '[${value.map(canonicalJson).join(',')}]';
  if (value is Map) {
    final keys = value.keys.cast<String>().toList()..sort();
    return '{${keys.map((k) => '${jsonEncode(k)}:${canonicalJson(value[k])}').join(',')}}';
  }
  throw FormatException(
    'Unsupported canonical JSON value: ${value.runtimeType}',
  );
}

String sha256Hex(String text) => sha256.convert(utf8.encode(text)).toString();

/// Play Integrity `requestHash` binding a token to this account, claim,
/// receipt and single-use nonce.
String referralIntegrityRequestHash({
  required String accountId,
  required String claimId,
  required String nonce,
  required String receiptDigest,
}) => sha256Hex(
  canonicalJson({
    'account_id': accountId,
    'campaign_id': referralCampaignId,
    'claim_id': claimId,
    'nonce': nonce,
    'receipt_digest': receiptDigest,
  }),
);
